### DISCUSSION TITLE
Perfomance problem when Loading Workflow Definitions

### DISCUSSION THREAD

**Discussion by caiohomem** [elsa-workflows/elsa-core#1153] [Design & tech]
Hi,

I´m working with version 1.5.5

I have a problem of performance  when Loading Workflow Definitions using EF Core Provider for the cache and sometimes I have time out and causes Elsa not to start.

I have 10000 rows in the tables WorkflowDefinitionVersions, ActivityDefinitions, ConnectionDefintions.

This is the actual code.

            var query = dbContext
                .WorkflowDefinitionVersions
                .Include(x => x.Activities)
                .Include(x => x.Connections)
                .AsQueryable()
                .WithVersion(version);

These does a full join of the 3 tables on SQL and returns 170k rows.
This is the query generated by EF Core

SELECT [w].[Id], [w].[DefinitionId], [w].[Description], [w].[IsDisabled], [w].[IsLatest], [w].[IsPublished], [w].[IsSingleton], [w].[Name], [w].[Variables], [w].[Version], [w].[VersionId], [a].[Id], [a].[ActivityId], [a].[Description], [a].[DisplayName], [a].[Left], [a].[Name], [a].[State], [a].[Top], [a].[Type], [a].[WorkflowDefinitionVersionId], [c].[Id], [c].[DestinationActivityId], [c].[Outcome], [c].[SourceActivityId], [c].[WorkflowDefinitionVersionId]
FROM [WorkflowDefinitionVersions] AS [w]
LEFT JOIN [ActivityDefinitions] AS [a] 
ON [w].[Id] = [a].[WorkflowDefinitionVersionId]
LEFT JOIN [ConnectionDefinitions] AS [c] 
ON [w].[Id] = [c].[WorkflowDefinitionVersionId]
ORDER BY [w].[Version] DESC, [w].[Id], [a].[Id], [c].[Id]

I rewrote the code for the maximum performance. This code does 3 selects one for each table.

class EntityFrameworkCoreWorkflowDefinitionStore

        public async Task<IEnumerable<WorkflowDefinitionVersion>> ListAsync(
            VersionOptions version,
            CancellationToken cancellationToken = default)
        {
            var defintions = await dbContext
                .WorkflowDefinitionVersions
                .AsNoTracking()
                .WithVersion(version)
                .ToListAsync(cancellationToken);

            var actvities = await dbContext
                .ActivityDefinitions
                .AsNoTracking()
                .Where(x => defintions.Select(x => x.Id).Contains(x.WorkflowDefinitionVersionId))
                .ToListAsync();

            foreach (var defintion in defintions)
            {
                defintion.Activities = actvities.Where(x => x.WorkflowDefinitionVersionId == defintion.Id).ToList();
            }

            var connections = await dbContext
                .ConnectionDefinitions
                .AsNoTracking()
                .Where(x => defintions.Select(x => x.Id).Contains(x.WorkflowDefinitionVersionId))
                .ToListAsync();

            foreach (var defintion in defintions)
            {
                defintion.Connections = connections.Where(x => x.WorkflowDefinitionVersionId == defintion.Id).ToList();
            }

            return mapper.Map<IEnumerable<WorkflowDefinitionVersion>>(defintions);
        }

Those are the 3 selects generated with my code.

SELECT [w].[Id], [w].[DefinitionId], [w].[Description], [w].[IsDisabled], [w].[IsLatest], [w].[IsPublished], [w].[IsSingleton], [w].[Name], [w].[Variables], [w].[Version], [w].[VersionId]FROM [WorkflowDefinitionVersions] AS [w]ORDER BY [w].[Version] DESC

SELECT [a].[Id], [a].[ActivityId], [a].[Description], [a].[DisplayName], [a].[Left], [a].[Name], [a].[State], [a].[Top], [a].[Type], [a].[WorkflowDefinitionVersionId]FROM [ActivityDefinitions] AS [a]WHERE [a].[WorkflowDefinitionVersionId] IN (127, 115, 114, 110, 129, 128, 87, 109, 108, 86, 126, 449, 448, 125, 445, 123, 104, 85, 84, 88, 103, 124, 121, 113, 444, 447, 446, 112, 122, 65, 70, 73, 74, 75, 80, 81, 82, 83)

SELECT [c].[Id], [c].[DestinationActivityId], [c].[Outcome], [c].[SourceActivityId], [c].[WorkflowDefinitionVersionId]FROM [ConnectionDefinitions] AS [c]WHERE [c].[WorkflowDefinitionVersionId] IN (127, 115, 114, 110, 129, 128, 87, 109, 108, 86, 126, 449, 448, 125, 445, 123, 104, 85, 84, 88, 103, 124, 121, 113, 444, 447, 446, 112, 122, 65, 70, 73, 74, 75, 80, 81, 82, 83)

What´s your opinion?
If it´s Ok I want to submit a Pull Request with this change.

Best regards,
Caio

**Comment by mohdali** [comment]
We faced similar issues with Elsa 1 and ended up writing our custom EF store and use it instead. You will have more flexibility this way I think.



https://elsa-workflows.github.io/elsa-core/docs/extensibility/extensibility-persistence

**Comment by sfmskywalker** [comment]
That's a nice improvement, but better yet is the new EF Core provider shipping with Elsa 2, where workflow definition data (including activities and connections) are simply stored as JSON data in a column of the workflow definition record.

Since Elsa 1.5 is deprecated, I recommend upgrading to Elsa 2.0 if you can. If you can't, then I recommend following @mohdali 's suggestion.
